### PR TITLE
FEAT: /post commands

### LINKED ISSUE #1089: [FEAT] - Add `post` commands
### What feature would you like to see?

We want to add a couple of commands that allow server managers to create a post (such as a rules post) through Becca.

The purpose of this is to allow server managers to edit that post, rather than having a user post it and have the user be the only one who can edit it.

We'll need a `/post create` command, which takes a channel parameter and the message to post, and a `/post edit` command which takes the message link to edit and the new content.

This is unfortunately blocked, until Discord completes the slate2 update and allows multi-line inputs in slash commands.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Vyvy-vi**
# Pull Request

## Description

This PR adds `/post create`, `/post edit`, `/post delete` commands and an even listener for modal submissions.

## Related Issue

Closes #1089 

## Documentation

For _any_ version updates, please verify if the [documentation page](https://docs.beccalyria.com?utm_source=github&utm_medium=pr-template) needs an update. If it does, please [create an issue there](https://github.com/BeccaLyria/discord-documentation/issues/new?assignees=nhcarrigan&labels=%F0%9F%9A%A6+status%3A+awaiting+triage&template=update.md&title=%5BUPDATE%5D) to ensure it is not forgotten.

- [ ] My contribution does NOT require a documentation update.
- [x] My contribution DOES require a documentation update.

### CODE REVIEW

**Comment by Vyvy-vi** [comment]
There are a few open questions I have related to this PR:
- Q) How do we prevent the user from deleting or updating messages sent by Becca that weren't created using the `/post create` command? Eg: the reaction role message. Could we perhaps store an object in the DB that contains the IDs of post messages?
- Q) Some of the methods used to resolve channels and messages throw an error when they're not found. Should we add an extra try-catch block for this, or is there some other way of checking this so that we can send detailed error messages to the user?
- Q) Is there a way to fetch a Message with just the messageId without filtering through all the channels in a guild? (The modal generated by the `/post edit` command has a customId that has format - `pe-${channelId}-${messageId}`). Also, should these IDs be in cammelCase?

**Comment by naomi-lgbt** [comment]
> Q) How do we prevent the user from deleting or updating messages sent by Becca that weren't created using the /post create command? Eg: the reaction role message. Could we perhaps store an object in the DB that contains the IDs of post messages?

Given that this is locked to the Manage Guild permission, I wouldn't worry too much about this.

> Q) Some of the methods used to resolve channels and messages throw an error when they're not found. Should we add an extra try-catch block for this, or is there some other way of checking this so that we can send detailed error messages to the user?

For the `.fetch()` calls, we can chain a `.catch(()=>null)` and then assert for existence. If it doesn't exist, we can respond with a message instead of an error.

Q) Is there a way to fetch a Message with just the messageId without filtering through all the channels in a guild? (The modal generated by the /post edit command has a customId that has format - pe-${channelId}-${messageId}).

No, but you should be able to resolve the channel and then resolve the message from the channel.

Q) Since there is a 17.1% duplication in the PR ( aaaa sonarcloud isn't happy :( ), would it be alright if I extract some of the permission checks into the run method in post.js or add a function in utils to do those checks and use that function in all the files?

Don't worry too much about the duplication that's not a concern for us at this time.

**Comment by naomi-lgbt** [comment]
@Vyvy-vi is this one ready for review?

**Comment by Vyvy-vi** [comment]
This needs addition of .catch() for error handling in some parts. Maybe it can be reviewed once those changes have been commited?

**Comment by naomi-lgbt** [comment]
Sure thing~! Just let me know when~!

**Comment by naomi-lgbt** [comment]
Is this ready now?

**Comment by Vyvy-vi** [comment]
yes, it's ready!

File: src/commands/post.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+            .setName("link")
+            .setDescription("Link to the post that you want to edit.")
+            .setRequired(true)
+        )
+    )
+    .addSubcommand(
+      new SlashCommandSubcommandBuilder()
+        .setName("delete")
+        .setDescription("Command to delete a server post.")
+        .addIntegerOption((option) =>
```
Comment (naomi-lgbt) [inline src/commands/post.ts:62]:
```suggestion
        .addStringOption((option) =>
```

File: src/commands/post.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        .addChannelOption((option) =>
+          option
+            .setName("channel")
+            .setDescription("The channel to send this message in.")
+            .addChannelTypes(
+              ChannelType.GuildText,
+              ChannelType.GuildNews,
+              ChannelType.GuildNewsThread,
+              ChannelType.GuildPublicThread,
+              ChannelType.GuildPrivateThread
```
Comment (naomi-lgbt) [inline src/commands/post.ts:42]:
```suggestion
              ChannelType.GuildText,
              ChannelType.GuildAnnouncement,
              ChannelType.AnnouncementThread,
              ChannelType.PublicThread,
              ChannelType.PrivateThread,
```
Replace with the newer type names.

File: src/commands/subcommands/post/handleDelete.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    if (targetMessage.author !== Becca.user || !targetMessage.deletable) {
+      await interaction.reply({
+        content: t("commands:post.delete.cant-delete"),
+      });
+      return;
+    }
+    targetMessage.delete();
+    await interaction.reply({ content: t("commands:post.delete.success") });
```
Comment (naomi-lgbt) [inline src/commands/subcommands/post/handleDelete.ts:64]:
```suggestion
    await interaction.deferReply({ ephemeral: true });
    const { member, guild } = interaction;

    if (!guild) {
      await interaction.editReply({
        content: getRandomValue(t("responses:missingGuild")),
      });
      return;
    }

    const link = interaction.options.getString("link", true).split("/");
    const channelId = link[link.length - 2];
    const messageId = link[link.length - 1];

    const channel = guild.channels.resolve(channelId);

    if (
      !member ||
      typeof member.permissions === "string" ||
      !member.permissions.has(PermissionFlagsBits.ManageGuild)
    ) {
      await interaction.editReply({
        content: getRandomValue(t("responses:noPermission")),
      });
      return;
    }
    if (!channel || !("messages" in channel)) {
      await interaction.reply({
        content: t("commands:post.delete.invalid"),
      });
      return;
    }

    const targetMessage = await channel.messages
      .fetch(messageId)
      .catch(() => null);

    if (!targetMessage) {
      await interaction.editReply({
        content: t("commands:post.delete.doesnt-exist"),
      });
      return;
    }

    if (targetMessage.author !== Becca.user || !targetMessage.deletable) {
      await interaction.editReply({
        content: t("commands:post.delete.cant-delete"),
      });
      return;
    }
    await targetMessage.delete();
    await interaction.reply({ content: t("commands:post.delete.success") });
```

Ephemeral response, and await delete call.

File: src/commands/subcommands/post/handleEdit.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+      });
+      return;
+    }
+
+    if (targetMessage.author !== Becca.user) {
+      await interaction.reply({
+        content: t("commands:post.edit.cant-edit"),
+      });
+      return;
+    }
```
Comment (naomi-lgbt) [inline src/commands/subcommands/post/handleEdit.ts:68]:
```suggestion
    if (!guild) {
      await interaction.reply({
        content: getRandomValue(t("responses:missingGuild")),
        ephemeral: true,
      });
      return;
    }

    const link = interaction.options.getString("link", true).split("/");
    const channelId = link[link.length - 2];
    const messageId = link[link.length - 1];

    const channel = guild.channels.resolve(channelId);

    if (
      !member ||
      typeof member.permissions === "string" ||
      !member.permissions.has(PermissionFlagsBits.ManageGuild)
    ) {
      await interaction.reply({
        content: getRandomValue(t("responses:noPermission")),
        ephemeral: true,
      });
      return;
    }
    if (!channel || !("messages" in channel)) {
      await interaction.reply({
        content: t("commands:post.edit.invalid"),
      });
      return;
    }

    const targetMessage = await channel.messages
      .fetch(messageId)
      .catch(() => null);

    if (!targetMessage) {
      await interaction.reply({
        content: t("commands:post.edit.doesnt-exist"),
        ephemeral: true,
      });
      return;
    }

    if (targetMessage.author !== Becca.user) {
      await interaction.reply({
        content: t("commands:post.edit.cant-edit"),
        ephemeral: true,
      });
      return;
    }
```

Ephemeral.

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: src/commands/post.ts (hunk 1 above)
Comment (Vyvy-vi) [reply src/commands/post.ts:62]: 🤦

File: src/commands/post.ts [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+        .addChannelOption((option) =>
+          option
+            .setName("channel")
+            .setDescription("The channel to send this message in.")
+            .addChannelTypes(
+              ChannelType.GuildText,
+              ChannelType.GuildAnnouncement,
+              ChannelType.AnnouncementThread,
+              ChannelType.PublicThread,
+              ChannelType.PrivateThread,
```
Comment (naomi-lgbt) [inline src/commands/post.ts:42]:
```suggestion
              ChannelType.PrivateThread
```
One day I'll make suggestions on GitHub and **not** screw up the linter lol

**Review by naomi-lgbt — CHANGES_REQUESTED** [review]

File: src/commands/post.ts (hunk 5 above)
Comment (Vyvy-vi) [reply src/commands/post.ts:42]: lol
🤔 I wonder if there can be a github action that lints suggestions

**Review by naomi-lgbt — APPROVED** [review]
Okay, this should be good to go.

**Comment by naomi-lgbt** [comment]
## Thank You

Thank you for your continued contribution to our project. We have reviewed your pull request and are happy to accept these changes.

Please continue to watch for issues labelled `help wanted`, as these will be additional opportunities to contribute.

You can also see all open issues [through our contributor tool](https://contribute.naomi.lgbt)

**Pull request merged**